unityID - hagrawa2


## 1) Degree Distribution

For the graphs provided to you, test and report which graphs are scalefree,
namely whose degree distribution follows a power law , at least asymptotically. That is, the fraction P(k) of
nodes in the network having k connections to other nodes goes for large values of k as

\begin{equation*}
\ P(k) \sim k^{(-\gamma)} 
\end{equation*}

where γ is a parameter whose value is typically in the range 2 < γ < 3, although occasionally it
may lie outside these bounds.

In [2]:
import glob
import pandas as pd
import numpy as np
from IPython.display import display, HTML

files = glob.glob('degree-outputs/*')

result = {
    'filename': [],
    'alpha': [],
    'scalefree': []
}

for f in files:
    name = f.split('/')[1]
    result['filename'].append(name)
    
    df = pd.read_csv(f)
    # Show DF for the random generated graphs     
    if len(name.split('.')) == 2:
        print(name)
        display(df)
    
    count = list(df['count'])
    degree = list(df['degree'])
    total_nodes = sum(count)
    fraction = [float(c)/total_nodes for c in count]
    slope, intercept = np.polyfit(np.log(degree), np.log(fraction), 1)
    
    result['alpha'].append(abs(slope))
    result['scalefree'].append('True' if abs(slope) > 1 and abs(slope) < 3.5 else 'False')
    
df = pd.DataFrame(data=result)
display(df)

gnm2.csv


Unnamed: 0  degree  count
0            0     200     36
1            1     201     41
2            2     202     42
3            3     203     31
4            4     204     29
5            5     205     30
6            6     206     21
7            7     207     22
8            8     208     26
9            9     209     21
10          10     210     15
11          11     211     24
12          12     212     29
13          13     213     21
14          14     214     12
15          15     215     10
16          16     216     13
17          17     217     14
18          18     218      9
19          19     219     10
20          20     220      8
21          21     221      9
22          22     222     11
23          23     223      8
24          24     224      4
25          25     225      2
26          26     226      2
27          27     227      3
28          28     228      1
29          29     229      3
..         ...     ...    ...
40          40     166      1
41          41     170      2
42          42     172      4
43          43     173      3
44          44     174      5
45          45     175      6
46          46     176      2
47          47     177      4
48          48     178      3
49          49     179      4
50          50     180     12
51          51     181      9
52          52     182     13
53          53     183     12
54          54     184     20
55          55     185     12
56          56     186     18
57          57     187     17
58          58     188     25
59          59     189     26
60          60     190     16
61          61     191     25
62          62     192     33
63          63     193     30
64          64     194     27
65          65     195     29
66          66     196     37
67          67     197     30
68          68     198     30
69          69     199     19

[70 rows x 3 columns]

gnm1.csv


Unnamed: 0  degree  count
0            0      12      3
1            1      13      2
2            2      14      5
3            3      15      4
4            4      16      7
5            5      17      8
6            6      18     10
7            7      19      5
8            8      20     10
9            9      21     10
10          10      22     12
11          11      23      6
12          12      24      4
13          13      25      6
14          14      26      3
15          15      27      1
16          16      29      2
17          17      31      1
18          18      33      1

gnp2.csv


Unnamed: 0  degree  count
0            0       6      1
1            1       7      2
2            2       8      2
3            3       9      5
4            4      10      7
5            5      11     19
6            6      12     48
7            7      13     60
8            8      14     73
9            9      15    101
10          10      16    130
11          11      17    151
12          12      18    170
13          13      19    173
14          14      20    180
15          15      21    170
16          16      22    151
17          17      23    124
18          18      24    109
19          19      25     94
20          20      26     70
21          21      27     57
22          22      28     32
23          23      29     25
24          24      30     19
25          25      31     11
26          26      32      8
27          27      33      3
28          28      34      1
29          29      35      1
30          30      36      3

gnp1.csv


Unnamed: 0  degree  count
0           0       1      4
1           1       2     12
2           2       3     15
3           3       4     12
4           4       5     18
5           5       6     19
6           6       7     11
7           7       8      5
8           8       9      4

filename     alpha scalefree
0     dblp.graph.large.csv  2.671850      True
1     dblp.graph.small.csv  1.141093      True
2                 gnm2.csv  0.864608     False
3                 gnm1.csv  1.093619      True
4  youtube.graph.small.csv  2.604728      True
5   amazon.graph.large.csv  2.548099      True
6                 gnp2.csv  0.448687     False
7                 gnp1.csv  0.063797     False
8  youtube.graph.large.csv  1.733190      True
9   amazon.graph.small.csv  1.678353      True

Answer the following questions:
1. Generate a few random graphs. You can do this using networkx’s random graph generators or GTGraph . Do the random graphs you tested appear to be scalefree?


|filename	|alpha	|scalefree
|-----------|-----------|------|
|gnm2.csv	|0.864608	|False |
|gnm1.csv	|1.093619	|True  |
|gnp2.csv	|0.448687	|False |
|gnp1.csv	|0.063797	|False |


2. Do the Stanford graphs provided to you appear to be scalefree?


|filename	                |alpha	    |scalefree
|---------------------------|-----------|---------|
|dblp.graph.large.csv	    |2.671850	|True|
|dblp.graph.small.csv	    |1.141093	|True|
|youtube.graph.small.csv	|2.604728	|True|
|amazon.graph.large.csv	    |2.548099	|True|
|youtube.graph.large.csv	|1.733190	|True|
|9amazon.graph.small.csv	|1.678353	|True|

## 2 - Centrality

Answer the following questions about the graph:

1. Rank the nodes from highest to lowest closeness centrality.

In [25]:
df = pd.read_csv('closeness.csv')
df

Unnamed: 0 id  closeness
0           0  F   0.071429
1           1  C   0.071429
2           2  H   0.066667
3           3  D   0.066667
4           4  B   0.058824
5           5  E   0.058824
6           6  A   0.055556
7           7  G   0.055556
8           8  I   0.047619
9           9  J   0.034483

2. Suppose we had some centralized data that would sit on one machine but would be
shared with all computers on the network. Which two machines would be the best
candidates to hold this data based on other machines having few hops to access this
data?
   * Ans : F and C as they have the most degree of closeness amongst all other computers

## 3 - Articulation
Answer the following questions:
1. In this example, which members should have been targeted to best disrupt communication in the organization?

In [27]:
df = pd.read_csv('graphframe_false.csv')
display(df)

Unnamed: 0                      id  articulation
0           0            Mohamed Atta             1
1           1          Usman Bandukra             1
2           2       Mamoun Darkazanli             1
3           3  Essid Sami Ben Khemais             1
4           4           Djamal Beghal             1
5           5           Nawaf Alhazmi             1
6           6             Raed Hijazi             1

The above members should be targetted to best disrupt the organization